packages/arb-bridge-eth/contracts/arch/Marshaling.sol

```diff
@@ -0,0 +1,255 @@
+        Value.Data memory stack = Value.newNone();
+        Value.Data[] memory vals = new Value.Data[](2);
+
+        for (uint256 i = 0; i < wholeChunkCount; i++) {
+            vals[0] = Value.newInt(data.toUint(startOffset + i * 32));
+            vals[1] = stack;
+            stack = Hashing.getTuplePreImage(vals);
+        }
+
+        if (dataLength % 32 != 0) {
+            uint256 lastVal = data.toUint(startOffset + dataLength - 32);
+            lastVal <<= (32 - (dataLength % 32)) * 8;
+            vals[0] = Value.newInt(lastVal);
+            vals[1] = stack;
+            stack = Hashing.getTuplePreImage(vals);
+        }
+
+        vals[0] = Value.newInt(dataLength);
+        vals[1] = stack;
+
+        return Hashing.getTuplePreImage(vals);
+    }
+
+    function bytestackToBytes(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            bool valid,
+            uint256 offset,
+            bytes memory byteData
+        )
+    {
+        offset = startOffset;
+        uint8 valType = uint8(data[offset]);
+        offset++;
+        if (valType != Value.tupleTypeCode() + 2) {
+            return (false, offset, byteData);
+        }
+
+        uint256 byteCount;
+        (valid, offset, byteCount) = deserializeCheckedInt(data, offset);
+        if (!valid) {
+            return (false, offset, byteData);
+        }
+        uint256 fullChunkCount = byteCount / 32;
+        uint256 partialChunkSize = byteCount % 32;
+        uint256 totalChunkCount = fullChunkCount +
+            (partialChunkSize > 0 ? 1 : 0);
+
+        bytes32[] memory fullChunks = new bytes32[](fullChunkCount);
+        bytes memory partialChunk = new bytes(partialChunkSize);
+
+        uint256 fullChunkIndex = 0;
+
+        for (uint256 i = 0; i < totalChunkCount; i++) {
+            valType = uint8(data[offset]);
+            offset++;
+            if (valType != Value.tupleTypeCode() + 2) {
+                return (false, offset, byteData);
+            }
+
+            uint256 nextChunk;
+            (valid, offset, nextChunk) = deserializeCheckedInt(data, offset);
+            if (!valid) {
+                return (false, offset, byteData);
+            }
+
+            if (i == 0 && partialChunkSize > 0) {
+                bytes32 chunkBytes = bytes32(nextChunk);
+                for (uint256 j = 0; j < partialChunkSize; j++) {
+                    partialChunk[j] = chunkBytes[j];
+                }
+            } else {
+                fullChunks[fullChunkCount - 1 - fullChunkIndex] = bytes32(
+                    nextChunk
+                );
+                fullChunkIndex++;
+            }
+        }
+
+        valType = uint8(data[offset]);
+        offset++;
+        if (valType != Value.tupleTypeCode()) {
+            return (false, offset, byteData);
+        }
+        return (true, offset, abi.encodePacked(fullChunks, partialChunk));
+    }
+}
```

Comment:
can we break up bytestackToBytes and bytesToBytestack into helper functions with good names or some comments to describe what its doing?

Reply:
I broke up `bytestackToBytes` a bit and I added a bunch of comments to both. Certainly open to suggestions on refactoring, but I can't think of anything further that makes sense. I did add a ton of comments to each method so that should hopefully help